File: cargo/tests/goodbye_world_tests.rs
use chrono::Local;
use colored::Colorize;

use cargo_hello::goodbye_world::{print_banner, print_goodbye};

fn format_error_message(expected: &str, actual: &str) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

#[test]
fn print_goodbye_without_name_test() {
    let args = vec![];
    let expected_output = "Goodbye World from Cargo !";
    let actual_output = print_goodbye(&args);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

#[test]
fn print_goodbye_with_name_test() {
    let args = vec!["", "Alter", "Sabeh"];
    let expected_output = "Goodbye Alter Sabeh from Cargo !";
    let actual_output = print_goodbye(&args);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

#[test]
fn print_banner_test() {
    let expected_output = r#"
    ▄████▄  ▄▄▄      ██▀███   ▄████ ▒█████
   ▒██▀ ▀█ ▒████▄   ▓██ ▒ ██▒██▒ ▀█▒██▒  ██▒
   ▒▓█    ▄▒██  ▀█▄ ▓██ ░▄█ ▒██░▄▄▄▒██░  ██▒
   ▒▓▓▄ ▄██░██▄▄▄▄██▒██▀▀█▄ ░▓█  ██▒██   ██░
   ▒ ▓███▀ ░▓█   ▓██░██▓ ▒██░▒▓███▀░ ████▓▒░
   ░ ░▒ ▒  ░▒▒   ▓▒█░ ▒▓ ░▒▓░░▒   ▒░ ▒░▒░▒░
     ░  ▒    ▒   ▒▒ ░ ░▒ ░ ▒░ ░   ░  ░ ▒ ▒░
   ░         ░   ▒    ░░   ░░ ░   ░░ ░ ░ ▒
   ░ ░           ░  ░  ░          ░    ░ ░
   ░
"#
    .to_string();
    let actual_output = print_banner();
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

File: csharp/tests/Goodbye.World.Tests/GoodbyeWorldTests.cs
namespace Prog.Goodbye.World.Tests;

[TestFixture]
public class GoodbyeWorldTests
{
    /// <summary> <h1><c>FormatErrorMessage</c></h1>
    /// <br/><br/>This method is used to format the error message.<br/><br/>
    ///
    /// <b>Arguments:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>expected: string</c> - The expected list of factors.
    /// </item>
    /// <item>
    /// <c>actual: string</c> - The actual list of factors.
    /// </item>
    /// </list>
    ///
    /// <b>Return:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>string</c> - The formatted error message.
    /// </item>
    /// </list>
    /// </summary>
    private static string FormatErrorMessage(string expected, string actual)
    {
        return $"\nTest run at: {DateTime.Now:O}\nExpected : {expected}\nActual   : {actual}\n";
    }

    /// <summary> <h1><c>PrintGoodbyeWithoutNameTest</c></h1>
    /// <br/><br/>This method is used to test the PrintGoodbye method of
    /// GoodbyeWorld without arguments.<br/><br/>
    /// </summary>
    [Test]
    public void PrintGoodbyeWithoutNameTest()
    {
        string[] args = [];
        string expectedOutput = "Goodbye World from C# !";
        string actualOutput = GoodbyeWorld.PrintGoodbye(args);
        string errorMessage = FormatErrorMessage(expectedOutput, actualOutput);
        Assert.That(expectedOutput, Is.EqualTo(actualOutput), errorMessage);
    }

    /// <summary> <h1><c>PrintGoodbyeWithNameTest</c></h1>
    /// <br/><br/>This method is used to test the PrintGoodbye method of
    /// GoodbyeWorld with arguments.<br/><br/>
    /// </summary>
    [Test]
    public void PrintGoodbyeWithNameTest()
    {
        string[] args = ["Alter", "Sabeh"];
        string expectedOutput = "Goodbye Alter Sabeh from C# !";
        string actualOutput = GoodbyeWorld.PrintGoodbye(args);
        string errorMessage = FormatErrorMessage(expectedOutput, actualOutput);
        Assert.That(expectedOutput, Is.EqualTo(actualOutput), errorMessage);
    }

    /// <summary> <h1><c>PrintBannerTest</c></h1>
    /// <br/><br/>This method is used to test the PrintBanner method of
    /// GoodbyeWorld.<br/><br/>
    [Test]
    public void PrintBannerTest()
    {
        string expectedOutput = @"
        ███▄    █▓████▄▄▄█████▓
        ██ ▀█   █▓█   ▓  ██▒ ▓▒
       ▓██  ▀█ ██▒███ ▒ ▓██░ ▒░
       ▓██▒  ▐▌██▒▓█  ░ ▓██▓ ░
   ██▓ ▒██░   ▓██░▒████▒▒██▒ ░
   ▒▓▒ ░ ▒░   ▒ ▒░░ ▒░ ░▒ ░░
   ░▒  ░ ░░   ░ ▒░░ ░  ░  ░
   ░      ░   ░ ░   ░   ░
    ░           ░   ░  ░
    ░
";
        string actualOutput = GoodbyeWorld.PrintBanner();
        string errorMessage = FormatErrorMessage(expectedOutput, actualOutput);
        Assert.That(actualOutput, Is.EqualTo(expectedOutput), errorMessage);
    }
}
